### PR TITLE
Fix SyncBatchNorm running var update issue

### PR DESCRIPTION
**Pull request by unlimblue**
## Fix https://github.com/pytorch/pytorch/issues/22192

+ change signature: `func: batch_norm_gather_stats(Tensor input, Tensor mean, Tensor invstd, Tensor? running_mean, Tensor? running_var, float momentum, float eps, Tensor counts) -> (Tensor, Tensor)`
+ change cuda & cuda head
```cuda
std::tuple<Tensor, Tensor> batch_norm_gather_stats_cuda(const Tensor& self, const Tensor& mean, const Tensor& invstd, const Tensor& running_mean,
                                                        const Tensor& running_var, double momentum, double epsilon, int64_t count) {
                                                        const Tensor& running_var, double momentum, double epsilon, const Tensor& counts)
```
+ change python interface 
```python
class SyncBatchNorm(Function):
    def forward(self, input, weight, bias, running_mean, running_var, eps, momentum, process_group, world_size):
        ...
```

### CODE REVIEW

**Comment by unlimblue** [comment]
@mrshenli [pytorch_xla_linux_trusty_py3_6_gcc5_4_build](https://circleci.com/gh/pytorch/pytorch/2083371?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) could retest again.

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -1381,7 +1381,7 @@
   dispatch:
     CUDA: batch_norm_elemt_cuda
 
-- func: batch_norm_gather_stats(Tensor input, Tensor mean, Tensor invstd, Tensor? running_mean, Tensor? running_var, float momentum, float eps, int count) -> (Tensor, Tensor)
+- func: batch_norm_gather_stats(Tensor input, Tensor mean, Tensor invstd, Tensor? running_mean, Tensor? running_var, float momentum, float eps, Tensor counts) -> (Tensor, Tensor)
```
[the remark this answers is not in the document]
Comment (unlimblue) [reply aten/src/ATen/native/native_functions.yaml]: Copy that, and is it possible change `const Tensor& counts` to `const ArrayRef<index_t> counts`? I don't know the best way of convert  `ArrayRef<index_t>` to a pytorch `Tensor` in c++.

[the remark this answers is not in the document]
Comment (unlimblue) [reply aten/src/ATen/native/native_functions.yaml]: Thanks, @hux999 is trying to fix `Tensor counts` issue.

[the remark this answers is not in the document]
Comment (mrshenli) [reply aten/src/ATen/native/native_functions.yaml]: > I don't know the best way of convert ArrayRef<index_t> to a pytorch Tensor in c++.

Can you try using the `from_blob` API (learnt from @yf225 ).
See [forum post](https://discuss.pytorch.org/t/can-i-initialize-tensor-from-std-vector-in-libtorch/33236/2)

**Comment by unlimblue** [comment]
We have found the reason, and trying to fix that.

**Comment by unlimblue** [comment]
@mrshenli are there some ci cache issue? 
```bash
Jul 01 06:22:58 ======================================================================
Jul 01 06:22:58 ERROR: test_clip_grad_norm (__main__.TestNN)
Jul 01 06:22:58 ----------------------------------------------------------------------
Jul 01 06:22:58 Traceback (most recent call last):
Jul 01 06:22:58   File "test_nn.py", line 1829, in test_clip_grad_norm
Jul 01 06:22:58     norm = clip_grad_norm_(l.parameters(), max_norm, norm_type=norm_type)
Jul 01 06:22:58   File "/opt/conda/lib/python3.6/site-packages/torch/nn/utils/clip_grad.py", line 36, in clip_grad_norm_
Jul 01 06:22:58     clip_coef = torch.tensor(max_norm, device=device) / (total_norm + 1e-6)
Jul 01 06:22:58 UnboundLocalError: local variable 'device' referenced before assignment
Jul 01 06:22:58 
Jul 01 06:22:58 ----------------------------------------------------------------------
```

**Comment by mrshenli** [comment]
@unlimblue that's not your fault. Other PR (e.g., #22392) also hit this error.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -1419,7 +1419,7 @@
   dispatch:
     CUDA: batch_norm_elemt_cuda
 
-- func: batch_norm_gather_stats(Tensor input, Tensor mean, Tensor invstd, Tensor? running_mean, Tensor? running_var, float momentum, float eps, int count) -> (Tensor, Tensor)
+- func: batch_norm_gather_stats(Tensor input, Tensor mean, Tensor invstd, Tensor? running_mean, Tensor? running_var, float momentum, float eps, int[] counts) -> (Tensor, Tensor)
```
Comment (mrshenli) [inline aten/src/ATen/native/native_functions.yaml]: I believe this is still BC-breaking. As mentioned above, can you try keep both the original one and the new one?

**Comment by SsnL** [comment]
There is an issue on the test error: https://github.com/pytorch/pytorch/issues/22052. Ignore it.

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (gchanan) [reply aten/src/ATen/native/native_functions.yaml]: @mrshenli you should probably also deprecate the old one so we can get rid of it in the future.

**Pull request closed without merging**